### PR TITLE
Add component instantiation validator

### LINKED ISSUE #227: Add error when components are instantiated outside of parent components
Because Needle needs to parse instantiations of `Component` to generate the DI graph, if a `Component` is instantiated outside of a parent `Component`, Needle may miss a path. A missed path will lead to a runtime error of missing dependency provider.

At the same time, simply parsing all source code is not enough to gather all DI paths, because SourceKit cannot provide all type information. For instance, the code below would require a global static analysis to retrieve the DI path:
```
func globalFunToCreateComponent(withParent parent: ComponentType) -> MyComponent {
    return MyComponent(parent: parent)
}
```
The actual type of the parent component of `MyComponent` is unknown until a global static analysis is performed. This would significantly violate Needle's performance requirements.

### PR DESCRIPTION
**Pull request by neakor**
`ComponentInstantiationValidator` validates that all `Component`s are instantiated within a parent `Component`. See issue for further details.

Fixes #227

### CODE REVIEW

File: Generator/Sources/NeedleFramework/Parsing/Processors/ComponentInstantiationValidator.swift [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                    // argument to check for cases such as `self.blah`.
+                    let argRange = NSRange(location: matchRange.location + matchRange.length, length: 5)
+                    let arg = content.substring(with: argRange)?.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
+                    // Special case for root component, where it is instantiated
+                    // with BootstrapComponent().
+                    let rootArgRange = NSRange(location: matchRange.location + matchRange.length, length: 20)
+                    let rootArg = content.substring(with: rootArgRange)?.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
+
+                    if let arg = arg, !validate(componentInstantiationArg: arg, and: rootArg) {
+                        throw GeneratorError.withMessage("\(componentName) is instantiated incorrectly. All components must be instantiated by parent components.")
```
Comment (rudro) [inline Generator/Sources/NeedleFramework/Parsing/Processors/ComponentInstantiationValidator.swift]: I would add something to the effect of "by passing self as the parent param" at the end of this.

File: Generator/Sources/NeedleFramework/Parsing/Processors/ComponentInstantiationValidator.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        }))
+        self.fileContents = fileContents
+    }
+
+    /// Process the data models.
+    ///
+    /// - throws: `ProcessingError` if any component instantiation is
+    /// invalid.
+    func process() throws {
+        for content in fileContents {
```
Comment (rudro) [inline Generator/Sources/NeedleFramework/Parsing/Processors/ComponentInstantiationValidator.swift]: So we're deciding not to multi-thread this ?

File: Generator/Sources/NeedleFramework/Parsing/FileFilters/ContentFilter.swift [hunk 3]
```diff
@@ -16,6 +16,10 @@
 
 import Foundation
 
+/// The regex pattern that matches any component instantiation expressions
+/// with the first capture group capturing the name of the component.
```
Comment (rudro) [inline Generator/Sources/NeedleFramework/Parsing/FileFilters/ContentFilter.swift]: Maybe add a note that this really catches the instantiation of any class where the constructor has `parent` as the first argument.

File: Generator/Sources/NeedleFramework/Parsing/Processors/ComponentInstantiationValidator.swift (hunk 1 above)
Comment (neakor) [reply Generator/Sources/NeedleFramework/Parsing/Processors/ComponentInstantiationValidator.swift]: Done.

**Pull request merged**